platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -2501,52 +2531,75 @@ namespace platoon_strategic_ihp
-        double currentGap = pm_.getDistanceToPredVehicle();  //TODO: fix this! it always returns 0.
+        // Note that pm_ only represents host's platoon members. As the aborting leader, the only vehicle
+        // preceding host is the candidate joiner. For this code to work, it depends on the candidate to publish
+        // mobility operation STATUS messages so that host can include it in the pm_ platoon membership.
+        double currentGap = pm_.getDistanceToPredVehicle();
         ROS_DEBUG_STREAM("Based on desired join time gap, the desired join distance gap is " << desiredJoinGap2 << " m");
         ROS_DEBUG_STREAM("Since we have max allowed gap as " << config_.desiredJoinGap << " m then max join gap became " << maxJoinGap << " m");
         ROS_DEBUG_STREAM("The current gap to joiner is " << currentGap << " m");
         ROS_DEBUG_STREAM("current_plan valid = " << pm_.current_plan.valid << ", isFirstLeaderAbortRequest = " << isFirstLeaderAbortRequest_);
 
-        // Check if gap is big enough and if current plan is timeout.
+        // NOTE: The front join depends upon the joiner to publish op STATUS messages with this platoon's ID, then host receives at least one
+        // and thereby adds the joiner to the platoon record. This process requires host's mob_req_cb_leader() to ACK the join request, then 
+        // the remote vehicle to handle that ACK and broadcast its first op STATUS, then host to receive and process it. All that has to happen
+        // before the below code block runs, even though this method starts to spin immediately after we send out the mentioned ACK. To avoid
+        // a race condition, we must wait a few cycles to ensure the 2-way messaging has completed.
+
+        // Check if gap is big enough and if there is no currently active plan and this method has been called several times
         // Add a condition to prevent sending repeated requests (Note: This is a same-lane maneuver, so no need to consider lower bound of joining gap.)
-        if (currentGap <= maxJoinGap  &&  !pm_.current_plan.valid  &&  isFirstLeaderAbortRequest_) 
+        ++numLeaderAbortingCalls_;
+        if (currentGap <= maxJoinGap  &&  !pm_.current_plan.valid  &&  numLeaderAbortingCalls_ > 4) 
         {
+            //TODO: this is just a test - remove when satisfied that this var doesn't need to be in the above if statement
+            ROS_DEBUG_STREAM("Gap size good & no valid plan. isFirstLeaderAbortRequest = " << isFirstLeaderAbortRequest_);
+
             // compose frontal joining plan, senderID is the old leader 
             cav_msgs::MobilityRequest request;
             std::string planId = boost::uuids::to_string(boost::uuids::random_generator()());
             long currentTime = ros::Time::now().toNSec() / 1000000;
             request.header.plan_id = planId;
-            request.header.recipient_id = pm_.targetLeaderId;
+            request.header.recipient_id = pm_.platoonLeaderID; //the new joiner
```

Comment:
isnt this the front join request on the joiner's side? so why is the recipient's id the new joiner??

Reply:
This is on the old leader vehicle (run_leader_aborting).  platoonLeaderID is set in mob_op_cb_leader() to represent the joiner.  This is correct.